core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -3376,6 +3406,110 @@ private void setAttributeFromEntry(SetAttributeEntry entry)
     return mBlockMaster.getWorkerInfoList();
   }
 
+  @Override
+  public void updateUfsMode(String ufsPath, UnderFileSystem.UfsMode ufsMode)
+      throws InvalidPathException {
+    try (JournalContext journalContext = createJournalContext()) {
+      updateUfsModeAndJournal(ufsPath, ufsMode, journalContext);
+    }
+  }
+
+  /**
+   * Update the ufs mode for path and create a journal entry.
+   *
+   * @param ufsPath the ufs path
+   * @param ufsMode the ufs mode
+   * @param journalContext the journal context
+   * @throws InvalidPathException if path is not used  by any mount point
+   */
+  private void updateUfsModeAndJournal(String ufsPath, UnderFileSystem.UfsMode ufsMode,
+      JournalContext journalContext) throws InvalidPathException {
+    updateUfsModeInternal(ufsPath, ufsMode);
+
+    // Journal
+    File.UfsMode ufsModeEntry;
+    switch (ufsMode) {
+      case NO_ACCESS:
+        ufsModeEntry = File.UfsMode.NO_ACCESS;
+        break;
+      case READ_ONLY:
+        ufsModeEntry = File.UfsMode.READ_ONLY;
+        break;
+      default:
+        ufsModeEntry = File.UfsMode.READ_WRITE;
+        break;
+    }
+    File.UpdateUfsModeEntry ufsEntry =
+        File.UpdateUfsModeEntry.newBuilder().setUfsPath(ufsPath).setUfsMode(ufsModeEntry).build();
+    journalContext.append(JournalEntry.newBuilder().setUpdateUfsMode(ufsEntry).build());
+  }
+
+  /**
+   * Update ufs mode from journal entry.
+   *
+   * @param entry the update ufs mode journal entry
+   * @throws InvalidPathException if the path is not used by any mount point
+   */
+  private void updateUfsModeFromEntry(File.UpdateUfsModeEntry entry)
+      throws InvalidPathException {
+    String ufsPath = entry.getUfsPath();
+    UnderFileSystem.UfsMode ufsMode;
+    switch (entry.getUfsMode()) {
+      case NO_ACCESS:
+        ufsMode = UnderFileSystem.UfsMode.NO_ACCESS;
+        break;
+      case READ_ONLY:
+        ufsMode = UnderFileSystem.UfsMode.READ_ONLY;
+        break;
+      default:
+        ufsMode = UnderFileSystem.UfsMode.READ_WRITE;
+        break;
+    }
+    updateUfsModeInternal(ufsPath, ufsMode);
+  }
+
+  private void updateUfsModeInternal(String ufsPath, UnderFileSystem.UfsMode ufsMode)
+      throws InvalidPathException {
+    mUfsManager.setUfsMode(ufsPath, ufsMode);
+  }
+
+  /**
+   * Check if the specified operation type is allowed to the ufs.
+   *
+   * @param alluxioPath the Alluxio path
+   * @param op the operation type
+   * @throws AccessControlException if the specified operation is not allowed
+   */
+  private void checkUfsMode(AlluxioURI alluxioPath, OperationType opType)
+      throws AccessControlException, InvalidPathException {
+    MountTable.Resolution resolution = mMountTable.resolve(alluxioPath);
+    UnderFileSystem ufs = resolution.getUfs();
+    UnderFileSystem.UfsMode ufsMode = ufs.getOperationMode(mUfsManager.getPhysicalUfsState());
```

Comment:
Instead of giving the manager state to the ufs, can we just get the physical path of the ufs, and ask the manager? It seems strange to give out the ufs manager state to UFS.

Reply:
by adding the `getPhysicalPath` api, we can only pass down state relevant to the UFS. maintaining the state at ufs manager allows us to journal and recover